### ISSUE TITLE
Device-agnostic code

### ISSUE THREAD

**Issue by Kaixhin** [pytorch/pytorch#3202]
A common issue - for both new and more seasoned users, including myself - is how to write nice device-agnostic code. There seems to be several potential solutions (using a global CUDA flag, always calling `torch.cuda.is_available()`, using `tensor.new()` etc.), used by various people in various bits of code. Ideally, we should have "best practices" documented somewhere, like [here](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/cuda.rst); or at least some place where we can link to.

**Comment by Kaixhin** [comment]
Alternative proposal - we provide a method that deals with the most common use cases (tensors, variables and networks). We could have something similar to Chainer's [`cuda.get_array_module()`](http://docs.chainer.org/en/stable/tutorial/gpu.html), something like [fakecuda](https://github.com/soumith/fakecuda), something like a [wrapper](https://gist.github.com/Kaixhin/8af7261dc790e8c899b4c2b2db97186e), etc. We're not going to be "write once, run everywhere", but we can probably catch most use cases.

**Pros:** It makes it *much* simpler for people writing CPU/1 GPU code (presumably a majority of users). Depending on the approach it'll probably be fine for multiple GPUs too.

**Cons:** An abstraction of this form isn't foolproof, but I can't immediately think of any major issues. Setting everything to CUDA by default is probably going to break things, but a wrapper, being more manual, can be used selectively enough.

**Comment by apaszke** [comment]
I agree we should figure out a way to make writing device-agnostic code easier. I don't think fakecuda is a good idea, it's too magical and has global side effects. Here's a list of options I see, along with some comments. In general I'm not 100% happy with any of those, but I don't have any better ideas right now.

#### Add device-aware dtype

Example:
```python
dtype = torch.cuda.float
torch.ones(5, 5, dtype=dtype) # Returns on the current GPU
```

I think that sooner or later we should add `dtype` arg to `torch` functions anyway, so it seems like a good enough extension.

#### `get_array_module()`

I think I'm fine with this option too, but I'm afraid you would either need a single global module captured inside functions, or use it everywhere inside. In general it can be simulated using the `dtype` approach using partials (of course this is an inefficient implementation, but you get the idea):

```python
import torch as _torch

dtype = ... # pick CUDA or CPU dtype
class DeviceAgnosticModule(object):
    def __getattribute__(self, name):
        obj = getattr(torch, name)
        if not callable(obj):
            return obj
        return functools.partial(obj, dtype=dtype)
```

#### `cast`

This is fine, except that you have to call it a lot and it makes the code super verbose. A similar alternative that people sometimes use now is this:
```python
dtype = torch.cuda.FloatTensor
x = Variable(torch.randn(5, 5).type(dtype))
```

#### `new`

This is really the best option, as it will give you a tensor of exactly the same type and on exactly the same device as the source. This is usually the desired behaviour, which is why we use it all over the place in the core library. On the other hand, it feels a bit unnatural, and is annoying when you do more complicated things, because you first construct the tensor and only then fill it with the data. `ones_like` and `zeros_like` helps a lot with that

**Comment by Kaixhin** [comment]
Thanks for more in-depth feedback on some options. For existing practices, highlighting `new`, `ones_like` and `zeros_like` are probably the way to go.

In terms of what to do when there aren't existing tensors, what I see as the main target is the [MNIST example](https://github.com/pytorch/examples/blob/master/mnist/main.py). Precisely, can we do [module casting](https://github.com/pytorch/examples/blob/master/mnist/main.py#L71-L72) and [tensor/variable casting](https://github.com/pytorch/examples/blob/master/mnist/main.py#L79-L80) without using the if statement? The cast/dtype solution seems good for tensors, but anything we can do for modules?

**Comment by Kaixhin** [comment]
I've put together some of this and other stuff I know into "best practices" documentation in https://github.com/pytorch/pytorch/pull/3227. I think we still want to think about future possibilities, but there's plenty that's worth noting down now for users.

**Comment by LMescheder** [comment]
It would indeed be amazing if PyTorch could make it simpler to write device agnostic code.

@apaszke I really like your first suggestion.
Just a quick note from my (applied) perspective: maybe it would be possible to split up the `dtype` and the `device` into two attributes? First of all, I think this would make it easier to handle cases where tensors are on different gpus and second of all, it would allow to write code like this:
``
torch.ones((5, 5), dtype=torch.float, device=x.device) # Returns on same device as x
``
Don't know how practical this is though.

**Comment by gchanan** [comment]
@LMescheder one thing I like about @apaszke's `Add device-aware dtype` suggestion is that
 it makes it explicit and obvious when something is happening on cuda, rather than requiring users to know that non-negative device numbers refer to cuda.

Maybe this is a bad idea, but what about also taking an optional `device` kw in every function that takes a `dtype` but to have both `cuda` and non-`cuda` dtypes?  So passing `dtype=torch.cuda.float` would give you a cuda float tensor on the current GPU, passing `dtype=torch.float` would give you a CPU float, and passing `dtype=torch.cuda.float, device=5` would give you a cuda tensor on GPU 5.  Obviously we'd need error checking to check things like non-`cuda` devices don't have non-negative `device`s specified (which is kind of silly because _if_ the device is specified it tells you everything), but it does seem to make the common case of not specifying a device more clear.

**Comment by elistevens** [comment]
Speaking only as an end-use with no particular insight into how difficult it would be to code:

These suggestions make me feel like end users will end up having to think more about device semantics, not less. I'd like to see the API move away from having the tensor type imply the storage location entirely. Have a context manager that sets cpu vs. cuda, default device, etc. Something like:

```
_tensorDefaults_stack = []
@contextmanager
def setTensorDefaults(location=torch.cpu_tensors, device=None, dtype=torch.float, pin_memory=None):
    assert location != torch.cuda_tensors or device is not None
    assert device is None or pin_memory is None
    _tensorDefaults_stack.append((location, device, dtype, pin_memory))
    yield
    _tensorDefaults_stack.pop()

with setTensorDefauts(location=torch.cuda_tensors, device=3):
    x = torch.ones((5, ,5))
    y = torch.Tensor([1,2,3,4,5]) # uses whatever is at the top of the _tensorDefaults_stack
```

Then once this exists, `DataLoader` class wraps its calls to the `Dataset` with an appropriate context, as do Modules. I think that this results in everything Just Working(tm).

**Comment by gchanan** [comment]
I'm not convinced that making it easy for end users to avoid thinking about device semantics is the correct goal.  The different devices have different performance characteristics, amount of memory/cache, incur data transfer costs/synchronizations, etc.  My worry is that by papering over these issues, we make it really easy to write non-performant code and really difficult to figure out why.  This discussion is similar to a discussion in the distributed systems community:  http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.41.7628 is a good discussion of why trying to unify all interfaces might not be the best idea (there are admittedly additional complexities in distributed computing like partial failures).

Anyway, that's why I like making it obvious when there are cuda storage / device transfers going on, because those are obvious places to look at performance.

**Comment by t-vi** [comment]
I'm not an expert on any of this, but about the only two things that bug me about using `new` is when I have to cast are
1. when I don't have a prototypical tensor at hand (e.g. in a module it feels unnatural to use `self.param.data.new(...)...` so it might be neat to have a "canonical tensor" in modules that reacts to .cuda() and whose `.new` is exposed as `module.newtensor` or so.
2. when I want a different type `my_float_tensor.new().long().resize(x,y).do_something` is just a bit much.

Other than that, I think using it is cool, I think that there is not much advantage in
`torch.ones((5, 5), dtype=torch.float, device=x.device) `
over
`torch.ones((5,5), out=x.new())`

But this is only from my limited experience.

**Comment by apaszke** [comment]
Regarding the "canonical tensor" it's hard, because the module can have a lot of different parameters that have different types and are on different devices.
I agree that `.new().long().resize(x, y)` chains are awful. I think `.new_long(x, y)` (and similar for all other types) would be a convenient helper.

**Comment by t-vi** [comment]
Oh yes, something like `.new_long(x,y)` would be cool (or `.new(x,y,dtype=long)`).

For the "canonical tensor", I imagine that, for basic applications, having something that "does the right thing" whether called before or after `module.cuda(device=d)` would be great. 
For more advanced applications, the user could create their own set of prototype tensors and just do `proto_gpu1.new(x,y)`, `proto_gpu2.new_long(z)` or so. But that is just pure imagination, as I don't have multiple GPUs. :slightly_smiling_face:

**Comment by Kaixhin** [comment]
@gchanan there are several potential goals here, and they are somewhat conflicting. One is to make CPU/GPU code seamless for end users, with just one flag somewhere. On the other hand, that's likely to results in less-performant code and harder debugging.

It seems like the conversation is going towards device-aware code, but making it a little less clunky in places. Sounds like `dtype` is a good step forward, and was probably in the pipeline anyway?

**Comment by kyunghyuncho** [comment]
a latecomer here: 

i agree that it's not a good idea to hide things behind the curtain from users. though, in many cases, it might be easier to to have a global behavior of ignoring, e.g., gpu (i.e., make .cuda() do nothing). could we at least give some kind of global switch allowing this behiavour?
